modules/gapi/test/own/rmat_tests.cpp

```diff
@@ -0,0 +1,178 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+#include "../test_precomp.hpp"
+#include <opencv2/gapi/own/rmat.hpp>
+
+namespace opencv_test
+{
+using GMatDesc = cv::GMatDesc;
+using RMat = cv::gapi::own::RMat;
+using HostMat = cv::gapi::own::HostMat;
+
+class RMatImplCpuWRef : public RMat::Impl
+{
+    cv::Mat& m_mat;
+
+    class HostMatCpu : public HostMat
+    {
+        cv::Mat& m_mat;
+    public:
+        HostMatCpu(cv::Mat& m) : m_mat(m) {}
+        virtual cv::gapi::own::Mat getMat() const override { return cv::to_own(m_mat); }
+        virtual void flush() const override {}
+    };
+public:
+    RMatImplCpuWRef(cv::Mat& m) : m_mat(m) {}
+
+    virtual RMat::HostMatP access(RMat::AccessType) const override
+    {
+        return std::make_shared<HostMatCpu>(m_mat);
+    }
+
+    virtual GMatDesc desc() const override { return cv::descr_of(m_mat); }
+};
+
+class RMatImplCpuWCopy : public RMat::Impl
+{
+    cv::Mat& m_mat;
+
+    class HostMatCpuWCopy : public HostMat
+    {
+        cv::Mat& m_mat;
+        cv::Mat  m_copy;
+    public:
+        HostMatCpuWCopy(cv::Mat& m) : m_mat(m), m_copy(m.clone()) {}
+        virtual cv::gapi::own::Mat getMat() const override { return cv::to_own(m_copy); }
+        virtual void flush() const override { m_copy.copyTo(m_mat); }
+    };
+public:
+    RMatImplCpuWCopy(cv::Mat& m) : m_mat(m) {}
+
+    virtual RMat::HostMatP access(RMat::AccessType) const override
+    {
+        return std::make_shared<HostMatCpuWCopy>(m_mat);
+    }
+
+    virtual GMatDesc desc() const override { return cv::descr_of(m_mat); }
+};
+
+TEST(RMat, Test)
+{
+    cv::Size sz{8, 8};
+    cv::Mat mat(sz, CV_8UC1);
+    cv::randu(mat, cv::Scalar::all(127), cv::Scalar::all(40));
+
+    RMat rmat(std::make_shared<RMatImplCpuWRef>(mat));
```

Comment:
The API to create and wrap external memory to RMat needs to be more friendly.

Reply:
I can propose only some template `make_rmat` which call `std::make_shared` inside:
```
auto rmat = cv::make_rmat<RMatImplCpuWRef>(mat);
```

Reply:
Just to clarify: there always user defined constructor for `Impl` class, so it's not required cv::Mat to be in the signature, it can be some another handle or so

Reply:
May be, but this `make_rmat` is STL-ish not OpenCV-ish :)

Reply:
... it applies to our existing code as well